Add --all(-generics|-monos) flags to monos command

userzimmermann opened the pull request:
After #118 the second PR for #109 

This time for the `monos` command, and with even three new flags :D to cover the 2-dimensional output of generics and their individual monomorphizations:

```
-a, --all
    --all-generics
    --all-monos
```

Only some more tests are needed to get it out of `WIP`... I'll hurry up ;)

data-pup commented:
Awesome! Feel free to ping me when you think this is ready for review. Love the idea of multiple flags!

userzimmermann commented:
@data-pup: ping :)

I added the remaining tests... Unfortunately, in the [`monos_all_monos`](https://github.com/rustwasm/twiggy/pull/120/files#diff-4a5fa82f4d97c6bee01d9d4a01a8e09f) test output for the `--all-monos` flag, there is no generic function with more individual monomorphizations than the default `max_monos = 10`. So the effect of `--max-monos` is not so visible, except that there are  `Σ [237 Total Rows]` instead of the `Σ [235 Total Rows]` of the default `monos` test output. So I was thinking about adding `--max-generics 11` to the test call to also get that `core::ptr::drop_in_place` generic listed, the only one with more than 10 monomorphizations: https://github.com/rustwasm/twiggy/pull/120/files#diff-b08b9997ee2e3dd09c6414a42623db04R51 But then I thought I'm thinking too much ;P Or what's your opinion about that?

And BTW: I also changed the existing `--max-monos` flag description a bit into `... for each listed generic ...` instead of just `... for each generic ...` To make it a bit clearer ;) https://github.com/rustwasm/twiggy/pull/120/files#diff-e6bf097a1c8cd4d9cfc57a634c427921R370

userzimmermann commented:
@data-pup: I tracked down the TravisCI issue... So this is now blocked by #122 :D

data-pup commented:
Nice work! I should have some time to review this tonight :+1:

userzimmermann commented on opt/definitions.rs:
```diff
@@ -403,13 +422,15 @@ impl Monos {
     pub fn max_generics(&self) -> u32 {
-        self.max_generics
+        if self.all_generics_and_monos || self.all_generics { u32::MAX }
+        else { self.max_generics }
```
@fitzgen 

> Let me guess: you've done a bunch of lisp/scheme in the past haven't you

And all of that in Emacs ofc! But would've never thought that this would be visible here :D You already know myself better than I do ;P

userzimmermann replied:
@fitzgen: Hmm... `cargo fmt` only did:

```diff
diff --git a/opt/opt.rs b/opt/opt.rs
index b1646da..6c0fc5c 100644
--- a/opt/opt.rs
+++ b/opt/opt.rs
@@ -1,10 +1,7 @@
 //! Options for running `twiggy`.

 #![deny(missing_debug_implementations)]
-#![cfg_attr(
-    feature = "wasm",
-    feature(use_extern_macros)
-)]
+#![cfg_attr(feature = "wasm", feature(use_extern_macros))]

 #[macro_use]
 extern crate cfg_if;
diff --git a/twiggy/tests/tests.rs b/twiggy/tests/tests.rs
index e3d8633..33302ce 100644
--- a/twiggy/tests/tests.rs
+++ b/twiggy/tests/tests.rs
@@ -455,7 +455,11 @@ test!(
     "2"
 );

-test!(garbage_wee_alloc_top_10, "garbage", "./fixtures/wee_alloc.wasm");
+test!(
+    garbage_wee_alloc_top_10,
+    "garbage",
+    "./fixtures/wee_alloc.wasm"
+);

 test!(
     garbage_wee_alloc_all,
```

Which is both not related to this PR :/ So somehow my lispy/schemey style seems to be fine ;)

Can 'cargo fmt` also be used for linting btw? Would be nice to see such style issues automatically in PRs... I could implement some `RustFmtBear` in https://github.com/coala/coala-bears and then we could add gitmate.io as additional CI service for doing automated code analysis... Besides other nice stuff ;)

userzimmermann replied:
Ah! `cargo fmt --all -- --check` can be used for linting :)

data-pup replied:
[clippy](https://github.com/rust-lang-nursery/rust-clippy) can also help with linting, if you're looking to check style details while you're working :+1:

data-pup replied:
Regarding the output of `cargo fmt` here: This confused me for a second as well, but it turns out the `fmt` command cannot see the `definitions.rs` file itself directly. This is related to the `build.rs` file in that part of the workspace.

For now, we can just format this conditional manually. Rust's style guidelines say we should organize if/else blocks like this:

```rust
if some_condition {
  // stuff
} else {
  // other stuff
}
```

Once this is done, I think this can be merged. Thanks for sticking this one out, and thanks for the help @userzimmermann!

data-pup replied:
As a sidenote: You *could* also run `rustfmt` against that file directly, but there are some additional changes that makes that might interfere with the way that the `build.rs` file removes unwanted attributes.

userzimmermann replied:
@data-pup @fitzgen: Running `rustfmt --check definitions.rs` indeed reveals a lot :D 

Also that I didn't stick to the rules in #118 ;P 

```diff
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 1:
-
 // Fun times ahead!
 //
 // Apparently, proc-macros don't play well with `cfg_attr` yet, and their
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 9:
 // It's terrible! But it works for now.

 /// Options for configuring `twiggy`.
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
-#[structopt(about = "\n`twiggy` is a code size profiler.\n\nIt analyzes a binary's call graph to answer questions like:\n\n* Why was this function included in the binary in the first place?\n\n* What is the retained size of this function? I.e. how much space\n  would be saved if I removed it and all the functions that become\n  dead code after its removal.\n\nUse `twiggy` to make your binaries slim!")]
+#[derive(Clone, Debug, StructOpt)]
+#[structopt(
+    about = "\n`twiggy` is a code size profiler.\n\nIt analyzes a binary's call graph to answer questions like:\n\n* Why was this function included in the binary in the first place?\n\n* What is the retained size of this function? I.e. how much space\n  would be saved if I removed it and all the functions that become\n  dead code after its removal.\n\nUse `twiggy` to make your binaries slim!"
+)]
 pub enum Options {
     /// List the top code size offenders in a binary.
     #[structopt(name = "top")]
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 39:
     /// Find and display code and data that is not transitively referenced by
     /// any exports or public functions.
     #[structopt(name = "garbage")]
-    Garbage(Garbage)
+    Garbage(Garbage),
 }

 /// List the top code size offenders in a binary.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 46:
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, StructOpt)]
 #[wasm_bindgen]
 pub struct Top {
     /// The path to the input binary to size profile.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 130:
 }

 /// Compute and display the dominator tree for a binary's call graph.
-#[derive(Clone, Debug, Default)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, Default, StructOpt)]
 #[wasm_bindgen]
 pub struct Dominators {
     /// The path to the input binary to size profile.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 214:

 /// Find and display the call paths to a function in the given binary's call
 /// graph.
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, StructOpt)]
 #[wasm_bindgen]
 pub struct Paths {
     /// The path to the input binary to size profile.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 336:

 /// List the generic function monomorphizations that are contributing to
 /// code bloat.
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, StructOpt)]
 #[wasm_bindgen]
 pub struct Monos {
     /// The path to the input binary to size profile.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 418:

     /// The maximum number of generics to list.
     pub fn max_generics(&self) -> u32 {
-        if self.all_generics_and_monos || self.all_generics { u32::MAX }
-        else { self.max_generics }
+        if self.all_generics_and_monos || self.all_generics {
+            u32::MAX
+        } else {
+            self.max_generics
+        }
     }

     /// The maximum number of individual monomorphizations to list for each
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 429:
     /// generic function.
     pub fn max_monos(&self) -> u32 {
-        if self.all_generics_and_monos || self.all_monos { u32::MAX }
-        else { self.max_monos }
+        if self.all_generics_and_monos || self.all_monos {
+            u32::MAX
+        } else {
+            self.max_monos
+        }
     }

     /// Set whether to hide individual monomorphizations and only show the
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 464:
 }

 /// Diff the old and new versions of a binary to see what sizes changed.
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, StructOpt)]
 #[wasm_bindgen]
 pub struct Diff {
     /// The path to the old version of the input binary.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 524:

 /// Find and display code and data that is not transitively referenced by any
 /// exports or public functions.
-#[derive(Clone, Debug)]
-#[derive(StructOpt)]
+#[derive(Clone, Debug, StructOpt)]
 #[wasm_bindgen]
 pub struct Garbage {
     /// The path to the input binary to size profile.
Diff in \\?\C:\Users\userz\Projects\rust\wasm\twiggy\opt\definitions.rs at line 576:

     /// The maximum number of items to display.
     pub fn max_items(&self) -> u32 {
-        if self.all_items { u32::MAX } else { self.max_items }
+        if self.all_items {
+            u32::MAX
+        } else {
+            self.max_items
+        }
     }

     /// Set the maximum number of items to display.
```

data-pup replied:
Awesome! I think the final step now is to take care of those `if`/`else` statements' formatting, and this can be merged :+1: The `#[derive(..)]` attributes can probably be left alone, I think they might play into how the `build.rs` file does things

userzimmermann replied:
@data-pup: Actually `build.rs` only removes either `#[wasm_bindgen]` or `#[structopt(...)]` lines depending on the target: 

https://github.com/rustwasm/twiggy/blob/26de56dc0cb0fb811e8e50f38fc2d8ab66982675/opt/build.rs#L20-L21

So merging the `#[derive(...)]` attributes would be actually fine... But ofc not as part of this PR ;) I will put that in another one :D And by simplifying `build.rs` a bit to not process `definitions.rs` line-wise, even the `#[structopt(...)]`s could be multi-lined... But that's ofc not even part of that other proposed PR ;P

data-pup replied:
Ah, good eye! Hadn't delved into that file too much quite yet, good to know :smile_cat: Feel free to ping me once the formatting changes are made and I can merge this. Looking forward to any other PR's too! :tada:

userzimmermann replied:
@data-pup 

> Ah, good eye!

Not good enough ;) I didn't see the `.case_insensitive(true)` for `RegexBuilder`... So the `#[derive(StructOpt)]` lines also get removed for writing `wasm.rs`... But anyway! All that can be optimized in `build.rs`... for sure... later :) To reach full `rustfmt` comformity in `definitions.rs` :D 

For now I made the 2 `if`/`else` format changes and re-pushed

data-pup commented:
Awesome, thanks @userzimmermann! :tada:
